File: NewsServer/app/src/main/java/com/dasbikash/news_server/database/NewsServerDatabase.java
/*
 * Copyright 2019 www.dasbikash.org. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.dasbikash.news_server.database;

import android.content.Context;

import com.dasbikash.news_server.database.daos.ArticleDao;
import com.dasbikash.news_server.database.daos.CountryDao;
import com.dasbikash.news_server.database.daos.LanguageDao;
import com.dasbikash.news_server.database.daos.NewsPaperDao;
import com.dasbikash.news_server.database.daos.PageDao;
import com.dasbikash.news_server.database.daos.PageGroupDao;
import com.dasbikash.news_server.database.daos.UserPreferenceDataDao;
import com.dasbikash.news_server.display_models.entity.Article;
import com.dasbikash.news_server.display_models.entity.ArticleVisitHistory;
import com.dasbikash.news_server.display_models.entity.Country;
import com.dasbikash.news_server.display_models.entity.Language;
import com.dasbikash.news_server.display_models.entity.Newspaper;
import com.dasbikash.news_server.display_models.entity.Page;
import com.dasbikash.news_server.display_models.entity.PageGroup;
import com.dasbikash.news_server.display_models.entity.UserPreferenceData;
import com.dasbikash.news_server.display_models.room_converters.DateConverter;
import com.dasbikash.news_server.display_models.room_converters.IntListConverter;
import com.dasbikash.news_server.display_models.room_converters.StringListConverter;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import androidx.room.TypeConverters;

@Database(entities = {Country.class,Language.class,Newspaper.class,
                        Page.class, PageGroup.class,Article.class,
        UserPreferenceData.class, ArticleVisitHistory.class},
        version = 1,exportSchema = false)
@TypeConverters({
        //ImageLinkListConverter.class,
        //IntDataListConverter.class,
        DateConverter.class,
        IntListConverter.class,
        StringListConverter.class
})
public abstract class NewsServerDatabase extends RoomDatabase {

    private static final String DATABASE_NAME = "news_server_database";

    private static volatile NewsServerDatabase INSTANCE;

    public static NewsServerDatabase getDatabase(final Context context) {
        if (INSTANCE == null) {
            synchronized (NewsServerDatabase.class) {
                if (INSTANCE == null) {
                    INSTANCE = Room.databaseBuilder(
                                context.getApplicationContext(),
                                NewsServerDatabase.class, DATABASE_NAME)
                            .build();
                }
            }
        }
        return INSTANCE;
    }

    public abstract CountryDao getCountryDao();
    public abstract LanguageDao getLanguageDao();
    public abstract NewsPaperDao getNewsPaperDao();
    public abstract PageDao getPageDao();
    public abstract PageGroupDao getPageGroupDao();
    public abstract ArticleDao getArticleDao();


    public abstract UserPreferenceDataDao getUserPreferenceDataDao();
}

File: NewsServer/news_server_data/src/main/java/com/dasbikash/news_server_data/database/NewsServerDatabase.kt
/*
 * Copyright 2019 das.bikash.dev@gmail.com. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.dasbikash.news_server_data.database

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.room.TypeConverters
import androidx.room.migration.Migration
import androidx.sqlite.db.SupportSQLiteDatabase
import com.dasbikash.news_server_data.data_sources.data_services.web_services.firebase.RealtimeDBUserSettingsUtils
import com.dasbikash.news_server_data.database.daos.*
import com.dasbikash.news_server_data.database.room_converters.ArticleImageConverter
import com.dasbikash.news_server_data.database.room_converters.DateConverter
import com.dasbikash.news_server_data.database.room_converters.IntListConverter
import com.dasbikash.news_server_data.database.room_converters.StringListConverter
import com.dasbikash.news_server_data.models.room_entity.*
import com.dasbikash.news_server_data.repositories.UserSettingsRepository

@Database(entities = [Country::class, Language::class, Newspaper::class, Page::class, Article::class,
                        ArticleVisitHistory::class,SavedArticle::class,ArticleSearchKeyWord::class,NewsCategory::class,
                        FavouritePageEntry::class],
        version = 3, exportSchema = false)
@TypeConverters(DateConverter::class, IntListConverter::class, StringListConverter::class, ArticleImageConverter::class)
internal abstract class NewsServerDatabase internal constructor(): RoomDatabase() {

    abstract val countryDao: CountryDao
    abstract val languageDao: LanguageDao
    abstract val newsPaperDao: NewsPaperDao
    abstract val pageDao: PageDao
    abstract val articleDao: ArticleDao
    abstract val savedArticleDao: SavedArticleDao
    abstract val articleSearchKeyWordDao: ArticleSearchKeyWordDao
    abstract val newsCategoryDao: NewsCategoryDao
    abstract val favouritePageEntryDao: FavouritePageEntryDao

    companion object {

        private val DATABASE_NAME = "news_server_database"

        @Volatile
        private lateinit var INSTANCE: NewsServerDatabase

        private val MIGRATION_1_2 = object : Migration(1, 2) {
            override fun migrate(database: SupportSQLiteDatabase) {
                database.execSQL("CREATE TABLE `ArticleSearchKeyWord` (`id` TEXT NOT NULL, `created` INTEGER NOT NULL, PRIMARY KEY(`id`))")
                database.execSQL("CREATE TABLE `NewsCategory` (`id` TEXT NOT NULL, `name` TEXT DEFAULT NULL, PRIMARY KEY(`id`))")
            }
        }

        private val MIGRATION_2_3 = object : Migration(2, 3) {
            override fun migrate(database: SupportSQLiteDatabase) {
                database.execSQL("DROP TABLE `UserPreferenceData`")
                database.execSQL("DROP TABLE `PageGroup`")
                database.execSQL("CREATE TABLE `FavouritePageEntry` (`pageId` TEXT NOT NULL, `subscribed` INTEGER NOT NULL DEFAULT 0, PRIMARY KEY(`pageId`))")
                RealtimeDBUserSettingsUtils.signOutUser()
            }
        }

        internal fun getDatabase(context: Context): NewsServerDatabase {
            if (!::INSTANCE.isInitialized) {
                synchronized(NewsServerDatabase::class.java) {
                    if (!::INSTANCE.isInitialized) {
                        INSTANCE = Room.databaseBuilder(context.applicationContext,
                                                    NewsServerDatabase::class.java, DATABASE_NAME)
                                                    .addMigrations(MIGRATION_1_2,MIGRATION_2_3)
                                                    .build()
                    }
                }
            }
            return INSTANCE
        }
    }

    fun nukeAppSettings(){
        articleDao.nukeTable()
        pageDao.nukeTable()
        newsPaperDao.nukeTable()
        countryDao.nukeTable()
        languageDao.nukeTable()
        newsCategoryDao.nukeTable()
    }

}
